ksqldb-engine/src/main/java/io/confluent/ksql/internal/UtilizationMetricsListener.java

```diff
@@ -0,0 +1,179 @@
+import org.slf4j.LoggerFactory;
+
+import java.io.BufferedReader;
+import java.io.FileReader;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+public class UtilizationMetricsListener implements Runnable, QueryEventListener {
+
+    private final String STREAM_THREAD_GROUP = "stream-thread-metrics";
+    private final String THREAD_ID = "thread-id";
+
+    private final List<KafkaStreams> kafkaStreams;
+    private final Logger logger = LoggerFactory.getLogger(UtilizationMetricsListener.class);
+    private final List<String> metrics;
+    private final Time time;
+    private final long windowSize;
+    private long lastSampleTime;
+
+    private final Map<String, Double> previousPollTime;
+    private final Map<String, Double> previousRestoreConsumerPollTime;
+    private final Map<String, Double> previousSendTime;
+    private final Map<String, Double> previousFlushTime;
+
+    public UtilizationMetricsListener(final long windowSize){
+        this.kafkaStreams = new ArrayList<>();
+        this.metrics = new LinkedList<>();
+        // we can add these here or pass it in through the constructor
+        metrics.add("poll-time-total");
+        metrics.add("restore-consumer-poll-time-total");
+        metrics.add("send-time-total");
+        metrics.add("flush-time-total");
+        time = Time.SYSTEM;
+        this.windowSize = windowSize;
+        lastSampleTime = time.milliseconds();
+        previousPollTime = new HashMap<>();
+        previousRestoreConsumerPollTime = new HashMap<>();
+        previousSendTime = new HashMap<>();
+        previousFlushTime = new HashMap<>();
+    }
+
+    // for testing
+    public UtilizationMetricsListener(final long windowSize, final List<KafkaStreams> streams, final Time time, final long lastSample) {
+        this.kafkaStreams = streams;
+        this.metrics = new LinkedList<>();
+        // we can add these here or pass it in through the constructor
+        metrics.add("poll-time-total");
+        metrics.add("restore-consumer-poll-time-total");
+        metrics.add("send-time-total");
+        metrics.add("flush-time-total");
+        this.time = time;
+        this.windowSize = windowSize;
+        lastSampleTime = lastSample;
+        previousPollTime = new HashMap<>();
+        previousRestoreConsumerPollTime = new HashMap<>();
+        previousSendTime = new HashMap<>();
+        previousFlushTime = new HashMap<>();
+    }
+
+    @Override
+    public void onCreate(
+            final ServiceContext serviceContext,
+            final MetaStore metaStore,
+            final QueryMetadata queryMetadata) {
+        kafkaStreams.add(queryMetadata.getKafkaStreams());
+    }
+
+    @Override
+    public void onDeregister(final QueryMetadata query) {
+        kafkaStreams.remove(query.getKafkaStreams());
+        previousPollTime.remove("poll-time-total");
```

Comment:
the keys being removed here should be thread ids right?

Reply:
Ah yeah, my bad. Did that without fully thinking. Just fixed